File: scripts/power_outages_duq.js
// scripts/power_outages_duq.js
import fs from "fs/promises";

const EVENTS_URL = "https://utilisocial.io/datacapable/v2/p/dlc/map/events";
const COUNTS_URL = "https://utilisocial.io/datacapable/v2/p/dlc/map/count?types=ZIP,COUNTY,MUNICIPALITY";

function getProp(props, key) {
  const found = (props || []).find((p) => p.property === key);
  return found ? found.value : [];
}

function toInt(value) {
  if (value === null || value === undefined || value === "") return 0;
  const n = Number(String(value).replace(/,/g, "").trim());
  return Number.isFinite(n) ? Math.trunc(n) : 0;
}

function normalizeName(value) {
  return String(value || "").trim().toUpperCase();
}

function calcPercent(out, served) {
  return served > 0 ? Number(((out / served) * 100).toFixed(2)) : null;
}

function extractGroupedCounts(payload) {
  // Supports a few possible response shapes.
  if (Array.isArray(payload)) return payload;
  if (Array.isArray(payload?.results)) return payload.results;
  if (Array.isArray(payload?.data)) return payload.data;
  if (Array.isArray(payload?.items)) return payload.items;
  return [];
}

function pickGroupType(row) {
  return String(
    row?.type ??
    row?.groupType ??
    row?.category ??
    row?.name ??
    ""
  ).toUpperCase();
}

function pickLabel(row) {
  return String(
    row?.label ??
    row?.key ??
    row?.value ??
    row?.name ??
    row?.title ??
    row?.group ??
    ""
  ).trim();
}

function pickCustomersOut(row) {
  return toInt(
    row?.customersOut ??
    row?.customers_out ??
    row?.numPeople ??
    row?.out ??
    row?.count
  );
}

function pickCustomersServed(row) {
  return toInt(
    row?.customersServed ??
    row?.customers_served ??
    row?.served
  );
}

async function fetchJson(url) {
  const res = await fetch(url, {
    headers: {
      "accept": "application/json, text/plain, */*",
      "user-agent": "PennAlerts Power Outages Duquesne Script"
    }
  });

  if (!res.ok) {
    throw new Error(`HTTP ${res.status} for ${url}`);
  }

  return res.json();
}

async function main() {
  try {
    const [eventsData, countsData] = await Promise.all([
      fetchJson(EVENTS_URL),
      fetchJson(COUNTS_URL)
    ]);

    const events = Array.isArray(eventsData) ? eventsData : [];

    const items = events.map((e) => {
      const props = e.additionalProperties || [];
      const counties = getProp(props, "counties");
      const municipalities = getProp(props, "municipalities");
      const zips = getProp(props, "zips");

      return {
        utility: "Duquesne Light",
        outages: toInt(e.numPeople),
        status: e.status || "Unknown",
        county: normalizeName(counties[0] || "UNKNOWN"),
        municipality: normalizeName(municipalities[0] || "UNKNOWN"),
        zip: String(zips[0] || "UNKNOWN").trim(),
        lat: typeof e.latitude === "number" ? e.latitude : null,
        lon: typeof e.longitude === "number" ? e.longitude : null
      };
    });

    const totalOutages = items.reduce((sum, item) => sum + toInt(item.outages), 0);

    // Municipalities from events
    const municipalitiesAgg = {};
    for (const item of items) {
      const muni = item.municipality || "UNKNOWN";
      municipalitiesAgg[muni] = (municipalitiesAgg[muni] || 0) + toInt(item.outages);
    }

    // Parse grouped counts
    const grouped = extractGroupedCounts(countsData);

    const countySummary = [];
    const countiesAgg = {};
    const zipSummary = [];
    const municipalitySummary = [];

    for (const row of grouped) {
      const type = pickGroupType(row);
      const label = normalizeName(pickLabel(row));
      const out = pickCustomersOut(row);
      const served = pickCustomersServed(row);

      if (!label) continue;

      if (type === "COUNTY") {
        countySummary.push({
          county: label,
          customers_out: out,
          customers_served: served,
          percent_out: calcPercent(out, served)
        });
        countiesAgg[label] = out;
      } else if (type === "ZIP") {
        zipSummary.push({
          zip: String(pickLabel(row)).trim(),
          customers_out: out,
          customers_served: served,
          percent_out: calcPercent(out, served)
        });
      } else if (type === "MUNICIPALITY") {
        municipalitySummary.push({
          municipality: label,
          customers_out: out,
          customers_served: served,
          percent_out: calcPercent(out, served)
        });
      }
    }

    // Fallback if grouped municipality counts are absent
    if (municipalitySummary.length === 0) {
      for (const [municipality, outages] of Object.entries(municipalitiesAgg)) {
        municipalitySummary.push({
          municipality,
          customers_out: outages,
          customers_served: null,
          percent_out: null
        });
      }
    }

    countySummary.sort((a, b) => {
      const diff = toInt(b.customers_out) - toInt(a.customers_out);
      if (diff !== 0) return diff;
      return a.county.localeCompare(b.county);
    });

    municipalitySummary.sort((a, b) => {
      const diff = toInt(b.customers_out) - toInt(a.customers_out);
      if (diff !== 0) return diff;
      return a.municipality.localeCompare(b.municipality);
    });

    zipSummary.sort((a, b) => {
      const diff = toInt(b.customers_out) - toInt(a.customers_out);
      if (diff !== 0) return diff;
      return String(a.zip).localeCompare(String(b.zip));
    });

    const output = {
      name: "power_outages_duq",
      utility: "Duquesne Light",
      fetched_at: new Date().toISOString(),
      total_outages: totalOutages,
      counties: countiesAgg,
      municipalities: municipalitiesAgg,
      county_summary: countySummary,
      municipality_summary: municipalitySummary,
      zip_summary: zipSummary,
      raw_count: items.length,
      items
    };

    await fs.mkdir("./data", { recursive: true });
    await fs.writeFile(
      "./data/power_outages_duq.json",
      JSON.stringify(output, null, 2),
      "utf8"
    );

    console.log("Wrote ./data/power_outages_duq.json");
    console.log(`Duquesne total outages: ${totalOutages}`);
    console.log(`County rows: ${countySummary.length}`);
    console.log(`Municipality rows: ${municipalitySummary.length}`);
    console.log(`ZIP rows: ${zipSummary.length}`);
  } catch (err) {
    console.error("Duquesne fetch error:", err);
    process.exit(1);
  }
}

main();

File: scripts/power_outages_duq.py
#!/usr/bin/env python3

import json
from pathlib import Path
from playwright.sync_api import sync_playwright


URL = "https://dlc.datacapable.com/map/?disableLinks=&utm_source=chatgpt.com"
OUTPUT_FILE = Path("data/power_outages_duq.json")


def to_int(value):
    try:
        if value is None:
            return 0
        return int(value)
    except Exception:
        try:
            return int(float(value))
        except Exception:
            return 0


def normalize_name(value):
    return str(value or "").strip().upper()


def calc_percent(out_val, served_val):
    out_num = to_int(out_val)
    served_num = to_int(served_val)
    if served_num <= 0:
        return None
    return round((out_num / served_num) * 100, 2)


def build_output(events_data, count_data):
    events = events_data if isinstance(events_data, list) else []
    counts = count_data if isinstance(count_data, list) else []

    items = []
    counties_from_items = {}
    municipalities_from_items = {}

    for e in events:
        props = e.get("additionalProperties") or []

        def get_prop(key):
            for p in props:
                if p.get("property") == key:
                    return p.get("value") or []
            return []

        counties = get_prop("counties")
        municipalities = get_prop("municipalities")
        zips = get_prop("zips")

        county = normalize_name(counties[0] if counties else "UNKNOWN")
        municipality = normalize_name(municipalities[0] if municipalities else "UNKNOWN")
        zip_code = str(zips[0] if zips else "UNKNOWN").strip()

        outages = to_int(e.get("numPeople"))

        item = {
            "utility": "Duquesne Light",
            "outages": outages,
            "status": e.get("status") or "Unknown",
            "county": county,
            "municipality": municipality,
            "zip": zip_code,
            "lat": e.get("latitude"),
            "lon": e.get("longitude"),
        }
        items.append(item)

        counties_from_items[county] = counties_from_items.get(county, 0) + outages
        municipalities_from_items[municipality] = municipalities_from_items.get(municipality, 0) + outages

    county_summary = []
    municipality_summary = []
    zip_summary = []

    for row in counts:
        row_type = str(row.get("type") or "").upper()
        name = str(row.get("name") or "").strip()
        customers_affected = to_int(row.get("customersAffected"))
        customers_served = to_int(row.get("customersServed"))

        if not name or not row_type:
            continue

        if row_type == "COUNTY":
            county_summary.append({
                "county": normalize_name(name),
                "customers_out": customers_affected,
                "customers_served": customers_served,
                "percent_out": calc_percent(customers_affected, customers_served),
            })
        elif row_type == "MUNICIPALITY":
            municipality_summary.append({
                "municipality": normalize_name(name),
                "customers_out": customers_affected,
                "customers_served": customers_served,
                "percent_out": calc_percent(customers_affected, customers_served),
            })
        elif row_type == "ZIP":
            zip_summary.append({
                "zip": name,
                "customers_out": customers_affected,
                "customers_served": customers_served,
                "percent_out": calc_percent(customers_affected, customers_served),
            })

    county_summary.sort(key=lambda x: (-to_int(x["customers_out"]), x["county"]))
    municipality_summary.sort(key=lambda x: (-to_int(x["customers_out"]), x["municipality"]))
    zip_summary.sort(key=lambda x: (-to_int(x["customers_out"]), x["zip"]))

    total_outages = sum(to_int(i.get("outages")) for i in items)

    output = {
        "name": "power_outages_duq",
        "utility": "Duquesne Light",
        "fetched_at": None,
        "total_outages": total_outages,
        "counties": counties_from_items,
        "municipalities": municipalities_from_items,
        "county_summary": county_summary,
        "municipality_summary": municipality_summary,
        "zip_summary": zip_summary,
        "raw_count": len(items),
        "items": items,
    }

    return output


def main():
    events_data = None
    count_data = None

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(
            viewport={"width": 1600, "height": 1000},
            user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36",
        )
        page = context.new_page()

        def handle_response(response):
            nonlocal events_data, count_data
            url = response.url

            try:
                if "/datacapable/v2/p/dlc/map/events" in url:
                    data = response.json()
                    if isinstance(data, list):
                        events_data = data

                if "/datacapable/v2/p/dlc/map/count?types=ZIP,COUNTY,MUNICIPALITY" in url:
                    data = response.json()
                    if isinstance(data, list):
                        count_data = data
            except Exception:
                pass

        page.on("response", handle_response)

        page.goto(URL, wait_until="networkidle", timeout=120000)

        try:
            page.locator("text=COUNTY").click(timeout=10000)
        except Exception:
            pass

       page.wait_for_response(
    lambda r: "count?types=ZIP,COUNTY,MUNICIPALITY" in r.url,
    timeout=15000
)

        if events_data is None:
            try:
                events_data = page.evaluate(
                    """
                    async () => {
                      const r = await fetch("https://utilisocial.io/datacapable/v2/p/dlc/map/events", {
                        credentials: "include"
                      });
                      return await r.json();
                    }
                    """
                )
            except Exception:
                events_data = []

        if count_data is None:
            try:
                count_data = page.evaluate(
                    """
                    async () => {
                      const r = await fetch("https://utilisocial.io/datacapable/v2/p/dlc/map/count?types=ZIP,COUNTY,MUNICIPALITY", {
                        credentials: "include"
                      });
                      return await r.json();
                    }
                    """
                )
            except Exception:
                count_data = []

        fetched_at = page.evaluate("() => new Date().toISOString()")

        browser.close()

    if not isinstance(events_data, list):
        raise RuntimeError("Failed to capture Duquesne events data")

    if not isinstance(count_data, list):
        raise RuntimeError("Failed to capture Duquesne count data")

    output = build_output(events_data, count_data)
    output["fetched_at"] = fetched_at

    OUTPUT_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(OUTPUT_FILE, "w", encoding="utf-8") as f:
        json.dump(output, f, indent=2)

    print(f"Wrote {OUTPUT_FILE}")
    print(f"Duquesne total outages: {output['total_outages']}")
    print(f"County rows: {len(output['county_summary'])}")
    print(f"Municipality rows: {len(output['municipality_summary'])}")
    print(f"ZIP rows: {len(output['zip_summary'])}")


if __name__ == "__main__":
    main()
